src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs

```diff
@@ -0,0 +1,1745 @@
+
+        internal int HashStrength
+        {
+            get
+            {
+                CheckThrow(true);
+                SslConnectionInfo info = Context.ConnectionInfo;
+                if (info == null)
+                {
+                    return 0;
+                }
+                return info.DataHashKeySize;
+            }
+        }
+
+        internal ExchangeAlgorithmType KeyExchangeAlgorithm
+        {
+            get
+            {
+                CheckThrow(true);
+                SslConnectionInfo info = Context.ConnectionInfo;
+                if (info == null)
+                {
+                    return (ExchangeAlgorithmType)0;
+                }
+                return (ExchangeAlgorithmType)info.KeyExchangeAlg;
+            }
+        }
+
+        internal int KeyExchangeStrength
+        {
+            get
+            {
+                CheckThrow(true);
+                SslConnectionInfo info = Context.ConnectionInfo;
+                if (info == null)
+                {
+                    return 0;
+                }
+                return info.KeyExchKeySize;
+            }
+        }
+
+        internal SslProtocols SslProtocol
+        {
+            get
+            {
+                CheckThrow(true);
+                SslConnectionInfo info = Context.ConnectionInfo;
+                if (info == null)
+                {
+                    return SslProtocols.None;
+                }
+
+                SslProtocols proto = (SslProtocols)info.Protocol;
+
+                // Restore client/server bits so the result maps exaclty on published constants.
+                if ((proto & SslProtocols.Ssl2) != 0)
+                {
+                    proto |= SslProtocols.Ssl2;
+                }
+                if ((proto & SslProtocols.Ssl3) != 0)
+                {
+                    proto |= SslProtocols.Ssl3;
+                }
+                if ((proto & SslProtocols.Tls) != 0)
+                {
+                    proto |= SslProtocols.Tls;
+                }
+                if ((proto & SslProtocols.Tls11) != 0)
+                {
+                    proto |= SslProtocols.Tls11;
+                }
+                if ((proto & SslProtocols.Tls12) != 0)
+                {
+                    proto |= SslProtocols.Tls12;
+                }
+                return proto;
+            }
+        }
+
+        internal Stream InnerStream
+        {
+            get
+            {
+                return _InnerStream;
+            }
+        }
+
+        internal StreamAsyncHelper InnerStreamAPM
+        {
+            get
+            {
+                return _InnerStreamAPM;
+            }
+        }
+
+        internal _SslStream SecureStream
+        {
+            get
+            {
+                CheckThrow(true);
+                if (_SecureStream == null)
+                {
+                    Interlocked.CompareExchange<_SslStream>(ref _SecureStream, new _SslStream(this), null);
```

Comment:
@bartonjs I'm not willing to risk behavior differences (if not bugs) changing this. 
I'm not exactly sure what parallel calls to Read or Write mean for the public SslStream given the sync implementation.